File: frontend/src/types/assignment.ts
export type AssignmentType =
  | 'WRITTEN'
  | 'PROBLEM_SOLVING'
  | 'PROGRAMMING'
  | 'PROJECT'
  | 'RESEARCH'
  | 'CREATIVE';

export type AssignmentDifficulty = 'EASY' | 'MEDIUM' | 'HARD';

export type AssignmentStatus = 'DRAFT' | 'ACTIVE' | 'PENDING_REVIEW' | 'COMPLETED';

export type SubmissionStatus =
  | 'NOT_STARTED'
  | 'DRAFT'
  | 'SUBMITTED'
  | 'LATE'
  | 'UNDER_REVIEW'
  | 'GRADED'
  | 'RETURNED'
  | 'RESUBMISSION_REQUIRED';

export type GradingMethod = 'MANUAL' | 'RUBRIC' | 'AI_ASSISTED';

export interface RubricCriterion {
  id?: number;
  criterion_name: string;
  max_points: number;
  description?: string;
  order_index?: number;
}

export interface SubmissionConfig {
  allowed_methods?: ('text' | 'file' | 'url')[];
  allowed_file_types?: string[];
  max_file_size_mb?: number;
  max_files_count?: number;
}

export interface GradingConfig {
  grading_method?: GradingMethod;
  enable_ai_assistance?: boolean;
  rubric_enabled?: boolean;
}

export interface TypeConfig {
  language?: string;
  allow_repo_url?: boolean;
  allow_zip?: boolean;
  required_concepts?: string[];
  step_by_step_required?: boolean;
  calculator_allowed?: boolean;
  min_word_count?: number;
  max_word_count?: number;
  citation_format?: string;
  required_sections?: string[];
  accepted_media_formats?: string[];
}

export interface Assignment {
  id: number;
  course_id: number;
  lesson_id?: number | null;
  teacher_id: number;
  title: string;
  description: string;
  instructions: string;
  assignment_type: AssignmentType;
  difficulty: AssignmentDifficulty;
  max_points: number;
  due_date?: string | null;
  status: AssignmentStatus | SubmissionStatus;
  submission_config?: SubmissionConfig | null;
  grading_config?: GradingConfig | null;
  type_config?: TypeConfig | null;
  created_at: string;
  updated_at: string;
  published_at?: string | null;

  course_title?: string | null;
  lesson_title?: string | null;
  total_submissions?: number;
  total_enrolled?: number;
  average_score?: number | null;
  pending_review_count?: number;
  rubric_criteria?: RubricCriterion[];
  my_submission?: Submission | null;
}

export interface SubmissionFile {
  id?: number | string;
  title?: string;
  filename: string;
  url: string;
  file_size?: number;
  mime_type?: string;
}

export interface Submission {
  id: number;
  assignment_id: number;
  student_id: number;
  status: SubmissionStatus;
  text_response?: string | null;
  external_url?: string | null;
  file_ids?: SubmissionFile[] | null;
  submitted_at?: string | null;
  created_at: string;
  updated_at: string;
  is_late: boolean;
  score?: number | null;
  feedback?: string | null;
  rubric_scores?: Record<string, number> | null;
  graded_at?: string | null;
  graded_by?: number | null;
  resubmission_reason?: string | null;

  student_name?: string | null;
  student_email?: string | null;
  student_avatar?: string | null;
  assignment_title?: string | null;
  assignment_max_points?: number | null;
  assignment_instructions?: string | null;
  assignment_type?: AssignmentType | null;
  rubric_criteria?: RubricCriterion[];
}

export interface AssignmentStats {
  total_assignments: number;
  active_assignments: number;
  pending_review_count: number;
  average_submission_rate: number;
  average_score: number;
}

export interface AIAssignmentGenRequest {
  course_id: number;
  lesson_id?: number | null;
  topic: string;
  assignment_type?: AssignmentType;
  difficulty?: AssignmentDifficulty;
  task_count?: number;
  custom_directives?: string;
  custom_prompt?: string;
}

export interface AIAssignmentGenResponse {
  title: string;
  description: string;
  instructions: string;
  tasks?: { title: string; description: string; points: number }[];
  assignment_type: AssignmentType;
  difficulty: AssignmentDifficulty;
  max_points: number;
  submission_config: SubmissionConfig;
  grading_config?: GradingConfig;
  type_config: TypeConfig;
  rubric_criteria: RubricCriterion[];
  optional_teacher_notes?: string;
}


export interface AIGradingAnalysisResponse {
  suggested_score: number;
  max_points: number;
  rubric_breakdown?: Record<string, number>;
  strengths: string[];
  areas_for_improvement: string[];
  suggested_feedback: string;
}

File: backend/app/schemas/assignment.py
from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, ConfigDict, Field

from app.enums.assignment import (
    AssignmentDifficulty,
    AssignmentStatus,
    AssignmentType,
    GradingMethod,
    SubmissionStatus,
)


# ----------------------------------------------------
# Rubric Criteria Schemas
# ----------------------------------------------------
class RubricCriterionCreate(BaseModel):
    criterion_name: str = Field(..., min_length=1)
    max_points: int = Field(..., gt=0)
    description: Optional[str] = None
    order_index: int = 0


class RubricCriterionResponse(RubricCriterionCreate):
    id: int
    assignment_id: int

    model_config = ConfigDict(from_attributes=True)


# ----------------------------------------------------
# Configuration Sub-Schemas
# ----------------------------------------------------
class SubmissionConfigSchema(BaseModel):
    allowed_methods: List[str] = Field(default_factory=lambda: ["file", "text"])  # "text", "file", "url"
    allowed_file_types: List[str] = Field(default_factory=lambda: ["pdf", "docx", "zip"])
    max_file_size_mb: int = 25
    max_files_count: int = 5


class GradingConfigSchema(BaseModel):
    grading_method: GradingMethod = GradingMethod.MANUAL
    enable_ai_assistance: bool = True
    rubric_enabled: bool = False


class TypeConfigSchema(BaseModel):
    # Programming
    language: Optional[str] = None
    allow_repo_url: Optional[bool] = None
    allow_zip: Optional[bool] = None
    required_concepts: Optional[List[str]] = None
    # Problem solving
    step_by_step_required: Optional[bool] = None
    calculator_allowed: Optional[bool] = None
    # Research
    min_word_count: Optional[int] = None
    max_word_count: Optional[int] = None
    citation_format: Optional[str] = None
    required_sections: Optional[List[str]] = None
    # Creative
    accepted_media_formats: Optional[List[str]] = None


# ----------------------------------------------------
# Assignment Base / Create / Update / Response
# ----------------------------------------------------
class AssignmentCreate(BaseModel):
    course_id: int
    lesson_id: Optional[int] = None
    title: str = Field(..., min_length=2, max_length=255)
    description: str
    instructions: str
    assignment_type: AssignmentType = AssignmentType.WRITTEN
    difficulty: AssignmentDifficulty = AssignmentDifficulty.MEDIUM
    max_points: int = Field(100, gt=0)
    due_date: Optional[datetime] = None
    status: AssignmentStatus = AssignmentStatus.DRAFT

    submission_config: Optional[Dict[str, Any]] = None
    grading_config: Optional[Dict[str, Any]] = None
    type_config: Optional[Dict[str, Any]] = None

    rubric_criteria: Optional[List[RubricCriterionCreate]] = None


class AssignmentUpdate(BaseModel):
    course_id: Optional[int] = None
    lesson_id: Optional[int] = None
    title: Optional[str] = None
    description: Optional[str] = None
    instructions: Optional[str] = None
    assignment_type: Optional[AssignmentType] = None
    difficulty: Optional[AssignmentDifficulty] = None
    max_points: Optional[int] = None
    due_date: Optional[datetime] = None
    status: Optional[AssignmentStatus] = None

    submission_config: Optional[Dict[str, Any]] = None
    grading_config: Optional[Dict[str, Any]] = None
    type_config: Optional[Dict[str, Any]] = None

    rubric_criteria: Optional[List[RubricCriterionCreate]] = None


class AssignmentResponse(BaseModel):
    id: int
    course_id: int
    lesson_id: Optional[int] = None
    teacher_id: int
    title: str
    description: str
    instructions: str
    assignment_type: AssignmentType
    difficulty: AssignmentDifficulty
    max_points: int
    due_date: Optional[datetime] = None
    status: Any

    submission_config: Optional[Dict[str, Any]] = None
    grading_config: Optional[Dict[str, Any]] = None
    type_config: Optional[Dict[str, Any]] = None


    created_at: datetime
    updated_at: datetime
    published_at: Optional[datetime] = None

    # Aggregated metrics for teacher list
    course_title: Optional[str] = None
    lesson_title: Optional[str] = None
    total_submissions: int = 0
    total_enrolled: int = 0
    average_score: Optional[float] = None
    pending_review_count: int = 0

    model_config = ConfigDict(from_attributes=True)


class AssignmentDetailResponse(AssignmentResponse):
    rubric_criteria: List[RubricCriterionResponse] = Field(default_factory=list)


class AssignmentStatsResponse(BaseModel):
    total_assignments: int = 0
    active_assignments: int = 0
    pending_review_count: int = 0
    average_submission_rate: float = 0.0
    average_score: float = 0.0


# ----------------------------------------------------
# Student Submission Schemas
# ----------------------------------------------------
class SubmissionCreate(BaseModel):
    text_response: Optional[str] = None
    external_url: Optional[str] = None
    file_ids: Optional[List[Dict[str, Any]]] = None  # e.g., [{"id": 1, "filename": "doc.pdf", "url": "..."}]
    is_draft: bool = False


class SubmissionGrade(BaseModel):
    score: float = Field(..., ge=0)
    feedback: Optional[str] = None
    rubric_scores: Optional[Dict[str, float]] = None


class SubmissionRequestResubmission(BaseModel):
    reason: str = Field(..., min_length=3)


class SubmissionResponse(BaseModel):
    id: int
    assignment_id: int
    student_id: int
    status: SubmissionStatus
    text_response: Optional[str] = None
    external_url: Optional[str] = None
    file_ids: Optional[List[Dict[str, Any]]] = None

    submitted_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime
    is_late: bool = False
    score: Optional[float] = None
    feedback: Optional[str] = None
    rubric_scores: Optional[Dict[str, float]] = None
    graded_at: Optional[datetime] = None
    graded_by: Optional[int] = None
    resubmission_reason: Optional[str] = None

    model_config = ConfigDict(from_attributes=True)


class SubmissionDetailResponse(SubmissionResponse):
    student_name: Optional[str] = None
    student_email: Optional[str] = None
    student_avatar: Optional[str] = None
    assignment_title: Optional[str] = None
    assignment_max_points: Optional[int] = None
    assignment_instructions: Optional[str] = None
    assignment_type: Optional[AssignmentType] = None
    rubric_criteria: List[RubricCriterionResponse] = Field(default_factory=list)


# ----------------------------------------------------
# AI Schemas
# ----------------------------------------------------
class AIAssignmentGenRequest(BaseModel):
    course_id: int
    lesson_id: Optional[int] = None
    topic: str
    assignment_type: AssignmentType = AssignmentType.WRITTEN
    difficulty: AssignmentDifficulty = AssignmentDifficulty.MEDIUM
    task_count: Optional[int] = Field(default=5, ge=1, le=20)
    custom_directives: Optional[str] = None
    custom_prompt: Optional[str] = None


class AIAssignmentGenResponse(BaseModel):
    title: str
    description: str
    instructions: str
    tasks: List[Dict[str, Any]] = Field(default_factory=list)
    assignment_type: AssignmentType
    difficulty: AssignmentDifficulty
    max_points: int
    submission_config: Dict[str, Any]
    grading_config: Optional[Dict[str, Any]] = None
    type_config: Optional[Dict[str, Any]] = None
    rubric_criteria: List[RubricCriterionCreate]
    optional_teacher_notes: Optional[str] = None



class AIGradingAnalysisResponse(BaseModel):
    suggested_score: float
    max_points: float
    rubric_breakdown: Optional[Dict[str, float]] = None
    strengths: List[str]
    areas_for_improvement: List[str]
    suggested_feedback: str
